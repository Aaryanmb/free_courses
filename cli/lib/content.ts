import { readFile, writeFile } from 'fs/promises';
import { COURSES_REGEX } from '../config/constants';

type ContentData = {
  title?: string;
  content: string;
};

export const buildContent = ({ title, content }: ContentData): string => {
  if (title) {
    return `### ${title}\n\n${content}`;
  }
  return content;
};

export const replaceContent = (original: string, data: string): string => {
  const newContent = original.replace(COURSES_REGEX, `$1\n\n${data}`);
  return newContent;
};

export const getFile = async (file: string): Promise<string | undefined> => {
  const data = await readFile(file, 'utf-8');
  if (!data) {
    throw new Error('No file found');
  }
  return data;
};

export const updateReadme = async (file: string, data: string): Promise<void> => {
  const readme = await getFile(file);
  if (!readme) {
    throw new Error('No readme found');
  }
  const newText = replaceContent(readme, data);
  await writeFile(file, newText, 'utf-8');
};
