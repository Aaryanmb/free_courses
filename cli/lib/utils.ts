export function isUrl(str: string): boolean {
  try {
    new URL(str);
    return true;
  } catch {
    return false;
  }
}

export function omit<T extends object, K extends keyof T>(obj: T, arr: K[]): Omit<T, K> {
  return Object.keys(obj)
    .filter((k) => !arr.includes(k as K))
    .reduce((acc, key) => {
      (acc as Record<string, unknown>)[key] = obj[key as K];
      return acc;
    }, {} as Partial<T>) as Omit<T, K>;
}

export function capitalizeFirstLetter(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function variableNameToSentence(str: string): string {
  return str
    .split(/(?=[A-Z])/)
    .map(capitalizeFirstLetter)
    .join(' ');
}

export function generateHyphens(length: number): string {
  return Array.from({ length }, () => '-').join('');
}

export function generateMarkdownUrl(displayText: string, url: string): string {
  return `[${displayText}](${url})`;
}
