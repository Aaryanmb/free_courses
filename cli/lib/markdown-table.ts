import { isUrl, variableNameToSentence, generateHyphens, generateMarkdownUrl } from './utils';

type MarkdownTableData = Record<string, unknown>[];

type MarkdownTableOptions = {
  urlKeyName?: string;
};

export function markdownTable(data: MarkdownTableData, { urlKeyName }: MarkdownTableOptions = {}): string {
  if (data.length === 0) {
    throw new Error('No data provided');
  }

  const headers = Object.keys(data[0]);
  const headerRow = `| ${headers.map(variableNameToSentence).join(' | ')} |`;
  const separatorRow = `| ${headers.map((header) => generateHyphens(header.length)).join(' | ')} |`;

  const rows = data
    .map((row) => {
      return `| ${headers
        .map((header) => {
          const value = row[header] ?? '';
          if (isUrl(value.toString())) {
            const displayText = urlKeyName ? row[urlKeyName]?.toString() : value;
            return generateMarkdownUrl(displayText as string, value as string);
          }
          return value;
        })
        .join(' | ')} |`;
    })
    .join('\n');

  return `${headerRow}\n${separatorRow}\n${rows}`;
}
