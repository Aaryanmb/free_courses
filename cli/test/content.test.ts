import * as fs from 'node:fs/promises';
import { beforeEach, assert, describe, expect, it, vi } from 'vitest';
import { buildContent, replaceContent, getFile, updateReadme } from '../lib/content';

beforeEach(() => {
  vi.resetAllMocks();
  vi.mock('node:fs/promises');
});

const originalContent = `# Hello World

## Courses

- [Course 1](https://example.com)
- [Course 2](https://example.com)
`;

const newContent = `- [Course 1](https://example.com)
- [Course 2](https://example.com)
- [Course 3](https://example.com)`;

describe('Content', () => {
  it('should build content with a title', () => {
    const content = buildContent({ title: 'Hello', content: 'World' });
    assert.equal(content, '### Hello\n\nWorld');
  });

  it('should build content without a title', () => {
    const content = buildContent({ content: 'Hello World' });
    assert.equal(content, 'Hello World');
  });

  it('should replace content', () => {
    const content = replaceContent(originalContent, newContent);
    expect(content).toMatchSnapshot();
  });
});

describe('getFile', () => {
  it('should return a file', async () => {
    vi.spyOn(fs, 'readFile').mockResolvedValue('file content');
    const file = await getFile('README.md');
    assert.equal(file, 'file content');
  });

  it('should return undefined if no file is found', async () => {
    vi.spyOn(fs, 'readFile').mockRejectedValueOnce('No file found');
    try {
      await getFile('README.md');
    } catch (error) {
      assert.equal(error, 'No file found');
    }
  });
});

describe('updateReadme', () => {
  it('should update the readme', async () => {
    vi.spyOn(fs, 'readFile').mockResolvedValue(originalContent);
    vi.spyOn(fs, 'writeFile').mockResolvedValue(undefined);
    await updateReadme('README.md', newContent);
    expect(fs.writeFile).toHaveBeenCalledWith('README.md', expect.any(String), 'utf-8');
  });

  it('should throw an error if no readme is found', async () => {
    vi.spyOn(fs, 'readFile').mockRejectedValueOnce('No file found');
    vi.spyOn(fs, 'writeFile').mockResolvedValue(undefined);
    try {
      await updateReadme('README.md', newContent);
    } catch (error) {
      assert.equal(error, 'No file found');
      expect(fs.writeFile).not.toHaveBeenCalled();
    }
  });
});
