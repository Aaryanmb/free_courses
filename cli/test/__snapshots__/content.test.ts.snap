// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Content > should replace content 1`] = `
"# Hello World

## Courses

- [Course 1](https://example.com)
- [Course 2](https://example.com)
- [Course 3](https://example.com)"
`;
